cp_diffdock_api.py

```diff
@@ -0,0 +1,261 @@
+        for ligand_i, ligand in enumerate(docking_request.ligands):
+            ligand_name, ligandData = ligand
+            ligand_label = f"ligand{ligand_i}"
+            sdf_file = work_dir / f"{ligand_label}.sdf"
+            try:
+                with open(sdf_file, 'w') as f:
+                    f.write(ligandData)
+                ligand_entries.append(PreppedComplexUnit(ligand_name, ligand_label, sdf_file))
+            except IOError as e:
+                msg = f"DiffDock failed to write ligand file: {e.strerror}"
+                ligand_entries.append(PreppedComplexUnit(ligand_name, error=msg))
+
+        # Write csv with docking entries for each protein+ligand complex
+        docking_entries: List[PreppedComplex] = []
+        for protein_entry in protein_entries:
+            for ligand_entry in ligand_entries:
+                complex_name = f"{protein_entry.label}-{ligand_entry.label}"
+                docking_entries.append(PreppedComplex(complex_name, protein_entry, ligand_entry))
+
+        csv_file = work_dir / "bmaps_diffdock.csv"
+        csv_content = DiffDockApi.make_request_csv(docking_entries)
+        if csv_content:
+            try:
+                with open(csv_file, 'w') as f:
+                    f.write(csv_content)
+            except IOError as e:
+                raise DiffDockSetupException(f"Could not write diffdock csv file: {e.strerror}")
+        else:
+            log.info(f"Couldn't prepare diffdock csv. Entries: {docking_entries}")
+            raise DiffDockSetupException("No suitable complexes")
+
+        out_dir = work_dir / "results"
+        return PreppedRequest(csv_file, out_dir, docking_entries)
+
+    @staticmethod
+    def process_results(preppedRequest: PreppedRequest) -> List[DiffDockProtocol.Response]:
+        """
+        DiffDock produces sdf files for each docked pose, in directories by complex name.
+        This function gathers the resulting sdfs for each complex to be returned by the DiffDock API.
+        """
+        allComplexResults = []
+
+        for complexInfo in preppedRequest.complex_infos:
+            # Find the output directory for this complex
+            complexName, proteinUnit, ligandUnit = complexInfo
+            proteinName = proteinUnit.name
+            ligandName = ligandUnit.name
+            resultDir = preppedRequest.out_dir / complexName
+            if not os.path.isdir(resultDir):
+                noPosesError = "No poses generated."
+                if not proteinUnit.ok():
+                    noPosesError += f" protein prep failure: {proteinUnit.error}."
+                if not ligandUnit.ok():
+                    noPosesError += f" ligand prep failure: {ligandUnit.error}."
+                allComplexResults.append(DiffDockProtocol.Result(proteinName, ligandName, error=noPosesError))
+                continue
+
+            # Process the sdfs in the output directory
+            poses = []
+            for dir_entry in os.scandir(resultDir):
```

Comment:
Instances of Path have .glob() and .iterdir() methods and .name attribute so you could ditch os.path.
Probably glob("*.sdf")